### PR TITLE
fix types

### PR DESCRIPTION
**Pull request by hipstersmoothie**
# What's Changing and Why

code review from @VojtechStep 

## What else might be affected

just types

## Tasks

-   [x] Add tests
-   [x] Update Documentation
-   [x] Update `jimp.d.ts`
-   [x] Add [SemVer](https://semver.org/) Label

### CODE REVIEW

**Comment by hipstersmoothie** [comment]
@VojtechStep any idea how to fix the buffer thing?

File: jimp.d.ts [hunk 1]
```diff
@@ -519,8 +519,8 @@ declare namespace Jimp {
             g: number,
             b: number,
             a: number,
-            cb: GenericCallback<number, T, Jimp>
-        ): this;
+            cb: GenericCallback<number, any, Jimp>
+        ): number;
```
Comment (VojtechStep) [inline jimp.d.ts]: This should actually return `void`

File: jimp.d.ts [hunk 2]
```diff
@@ -234,211 +234,211 @@ declare namespace Jimp {
...
+        colour(actions: ColorAction[], cb?: Jimp.ImageCallback): this;
 
         // Shape methods
-        rotate(deg: number, cb?: Jimp.ImageCallback<T>): this;
+        rotate(deg: number, cb?: Jimp.ImageCallback): this;
         rotate(
             deg: number,
-            mode: number | boolean | undefined | null,
-            cb?: Jimp.ImageCallback<T>
+            mode: number | boolean,
```
Comment (VojtechStep) [inline jimp.d.ts]: It seems I've made a mistake before, the type of mode should be `string | boolean` and it should be optional

File: jimp.d.ts [hunk 3]
```diff
@@ -447,42 +447,42 @@ declare namespace Jimp {
...
+        normalize(cb?: Jimp.ImageCallback): this;
+        invert(cb?: Jimp.ImageCallback): this;
+        gaussian(r: number, cb?: Jimp.ImageCallback): this;
         composite(
             src: Jimp,
             x: number,
             y: number,
-            options: Jimp.BlendModel,
-            cb?: Jimp.ImageCallback<T>
+            options: Jimp.BlendMode,
```
Comment (VojtechStep) [inline jimp.d.ts]: Since we provide a default value for `options`, it should be marked as optional

File: jimp.d.ts (hunk 1 above)
Comment (hipstersmoothie) [reply jimp.d.ts]: this one for sure returns a number though

Comment (VojtechStep) [reply jimp.d.ts]: Well [the function](https://github.com/oliver-moran/jimp/blob/master/src/index.js#L870) doesn't have a return statement

Comment (hipstersmoothie) [reply jimp.d.ts]: I think you're reading the diff wrong. line 523 is a part of `rgbaToInt` not `appendConstructorOption`

Comment (VojtechStep) [reply jimp.d.ts]: Oh 100% you're right, sorry about that 🤦‍♂️

**Comment by VojtechStep** [comment]
@hipstersmoothie Regarding the Buffer thing, I'm not sure how to approach it. We should add `@types/node` as a dependency of some sort, but there are drawbacks:
- Regular dependencies: won't go here, types are only for development
- DevDependencies: If we put it here, they are installed even for users of the library who don't use TypeScript
- OptionalDependencies: They still get installed
- PeerDependencies: Users don't need to install the types, but the will get annoying messages in the terminal if they don't

**Comment by hipstersmoothie** [comment]
Hmm. Better left to a different PR then. May just have to put something in the readme on how to configure it properly

**Comment by nrkn** [comment]
You forgot to make this callback optional:

```
        static rgbaToInt(
            r: number,
            g: number,
            b: number,
            a: number,
            cb: GenericCallback<number, any, Jimp>
        ): number;
```

**Comment by hipstersmoothie** [comment]
@nrkn please make a pr.

**Pull request merged**